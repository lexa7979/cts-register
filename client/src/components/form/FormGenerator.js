/**
 * MIT License
 *
 * Copyright (c) 2019 <alexander.urban@cygni.se>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/* eslint-disable class-methods-use-this */

import assert from "assert";

import React from "react";
import PropTypes from "prop-types";
import { Alert, Form, FormGroup, Col, Label, FormText } from "reactstrap";

import FieldTypes from "./field-types";

const propTypes = {
	fields:       PropTypes.object.isRequired,
	handleSubmit: PropTypes.func.isRequired,
	formClass:    PropTypes.string,
	validateData: PropTypes.func,
};

const defaultProps = {};

const initialState = {
	inputValues: {},
	touched:     {},
	messages:    {},
	actions:     [ "submit" ],
	alert:       null,
};

/**
 * Component which contains an input form
 */
export class FormGenerator extends React.Component {
	/**
	 * Initialising component
	 */
	constructor( props ) {
		super( props );

		this.state = { ...initialState };

		this.fieldList = [];
		for ( const name in this.props.fields ) {
			if ( typeof name === "string" && name !== "" ) {
				const { type = "input", value = "" } = this.props.fields[name];
				assert(
					FieldTypes[type] != null
					&& typeof FieldTypes[type] === "object"
					&& typeof FieldTypes[type].generateField === "function",
					`Unsupported field type "${type}" (${name})`
				);
				this.fieldList[name] = {
					...this.props.fields[name],

					name,
					type,
					id: `${this.props.formClass || "form"}-${name}`,
				};
				this.state.inputValues[name] = value;
			}
		}

		this.validation = {
			progressing: false,
			requested:   false,
		};

		this.mounted = false;
		this.alertTimeout = null;

		this.handleInputChange = this.handleInputChange.bind( this );
		this.runValidation = this.runValidation.bind( this );
		this.handleSubmit = this.handleSubmit.bind( this );
		this.handleReset = this.handleReset.bind( this );
	}

	/**
	 * Actions to take when component is ready
	 */
	componentDidMount() {
		this.mounted = true;
		this.runValidation();
	}

	/**
	 * Actions to take before component will be removed
	 */
	componentWillUnmount() {
		this.mounted = false;
	}

	/**
	 * Validates the form inputs if props.validateData is given
	 *
	 * If another (asynchronous) validation is already in progress,
	 * nothing will be done, now. Still, it's memorised that
	 * another validation shall be run just after the active one
	 * has finished.
	 */
	runValidation() {
		if ( typeof this.props.validateData !== "function" || !this.mounted ) {
			return Promise.resolve();
		}

		if ( this.validation.progressing ) {
			this.validation.requested = true;
			return Promise.resolve();
		}
		this.validation.progressing = true;

		return Promise.resolve( this.props.validateData( this.state.inputValues ) )
			.then( result => {
				this.validation.progressing = false;
				if ( this.validation.requested ) {
					this.validation.requested = false;
					this.runValidation();
					return;
				}

				if (
					this.mounted
					&& result != null && typeof result === "object"
					&& Array.isArray( result.actions )
					&& result.messages != null && typeof result.messages === "object"
				) {
					this.setState( { actions: result.actions, messages: result.messages } );
				}
			} );
	}

	/**
	 * Creates a React component
	 * if a global alert-message shall be shown.
	 *
	 * The information message automatically hide after some seconds.
	 *
	 * @returns	{object}
	 *		New React component representing the information-message.
	 */
	generateAlert() {
		if ( this.state.alert != null && this.alertTimeout == null ) {
			this.alertTimeout = setTimeout( () => {
				this.alertTimeout = null;
				if ( this.mounted ) {
					this.setState( { alert: null } );
				}
			}, 7000 );
		}

		return <Alert key="form-alert" color="success" isOpen={this.state.alert != null}>{this.state.alert}</Alert>;
	}

	/**
	 * Create a React component
	 * which represents a label related to the given input field.
	 *
	 * @param	{object}	fieldData
	 * @param	{string}	fieldData.name
	 * @param	{string}	fieldData.type
	 * @param	{string}	fieldData.id
	 * @param	{null|string}	fieldData.label
	 *
	 * @param	{object}	extraAttributes
	 *		Optional: Additional attributes which shall be included in the main-tag
	 *		of the new component.
	 *
	 * @returns	{object}
	 *		New React component
	 */
	generateLabel( fieldData, extraAttributes = {} ) {
		assert( fieldData != null && typeof fieldData === "object",
			"Invalid argument \"fieldData\"" );

		const { name, label, type, id } = fieldData;
		assert( typeof name === "string" && name !== "",
			`Invalid field property "name" (${name})` );
		assert( label == null || ( typeof label === "string" && label !== "" ),
			`Invalid field property "label" (${name}: ${label})` );
		assert( typeof type === "string" && type !== "",
			`Invalid field property "type" (${name}: ${type})` );
		assert( typeof id === "string" && id !== "",
			`Invalid field property "id" (${name}: ${id})` );

		const fieldProperties = {
			...extraAttributes,

			key:       `${name}-label`,
			className: `${label ? "label" : "nolabel"} ${type}`,
		};

		for ( const key in fieldProperties ) {
			if ( !fieldProperties[key] ) {
				delete fieldProperties[key];
			}
		}

		return label
			? React.createElement( Label, { ...fieldProperties, for: id }, label )
			: React.createElement( Col, fieldProperties );
	}

	/**
	 * Creates a React component
	 * if there is an error-message related to the input field with the given name.
	 *
	 * @param	{string}	fieldName
	 *
	 * @returns	{object|null}
	 *		New React component representing the error-message; or
	 *		Null if there is no related error.
	 */
	generateMessage( fieldName ) {
		assert( typeof fieldName === "string" && fieldName !== "", "Invalid argument" );

		return ( fieldName === "submit" || this.state.touched[fieldName] ) && this.state.messages[fieldName]
			? <FormText key={`${fieldName}-message`} color="danger">{this.state.messages[fieldName]}</FormText>
			: null;
	}

	/**
	 * Creates a React component
	 * which contains buttons to submit or reset the form input.
	 *
	 * @returns	{object}
	 *		New React component
	 */
	generateFinalRow() {
		const buttons = {};

		[ "submit", "reset" ].forEach( type => {
			const allTypeFields = Object.keys( this.fieldList )
				.filter( name => this.fieldList[name].type === type );
			const activeTypeFields = allTypeFields
				.filter( name => this.state.actions.includes( name ) );

			if ( activeTypeFields.length > 0 ) {
				buttons[type] = FieldTypes[type].generateField.call( this, this.fieldList[activeTypeFields[0]] );
			} else if ( allTypeFields.length > 0 ) {
				buttons[type] = FieldTypes[type].generateField.call( this, this.fieldList[allTypeFields[0]] );
			} else if ( type === "submit" ) {
				buttons[type] = FieldTypes[type].generateField.call( this, {} );
			}
		} );

		const submitMessage = this.generateMessage( "submit" );

		return <FormGroup key="form-final" row>
			<Col sm={{ size: 9, offset: 3 }}>
				{buttons.submit}{" "}{buttons.reset}
				{submitMessage}
			</Col>
		</FormGroup>;
	}

	/**
	 * Creates a React component
	 * which represent the input field with the given name
	 *
	 * @param	{string}	name
	 *
	 * @returns	{object}
	 *		New React component
	 */
	generateField( name ) {
		const data = this.fieldList[name];

		assert( data != null && typeof data === "object",
			`Invalid argument "name" (${name})` );
		assert( typeof data.type === "string" && typeof FieldTypes[data.type] === "object",
			`Invalid field type (${data.type})` );

		const id = `${this.props.formClass || "form"}-${data.name}`;

		const label = this.generateLabel( data, { sm: 3 } );
		const field = FieldTypes[data.type].generateField.call( this, data, { id } );
		const message = this.generateMessage( data.name );

		return <FormGroup row key={name}>
			{label}
			<Col sm={9}>
				{field}
				{message || ""}
			</Col>
		</FormGroup>;
	}

	/**
	 * Handles events
	 * whenever an input field of the form is changed
	 *
	 * @param	{object}	event
	 *		Description of the event to handle
	 */
	handleInputChange( event ) {
		const { name } = event.target;
		const value = event.target.type === "checkbox" ? event.target.checked : event.target.value;

		this.setState( { inputValues: { ...this.state.inputValues, [name]: value } }, () => {
			this.runValidation()
				.then( () => {
					if ( this.state.touched[name] !== true ) {
						this.setState( { touched: { ...this.state.touched, [name]: true } } );
					}
				} );
		} );
	}

	/**
	 * Handles events
	 * whenever a user clicks on one of the form's submit buttons
	 */
	handleSubmit( event = null ) {
		if ( event != null ) {
			event.preventDefault();
		}

		if ( typeof this.props.handleSubmit === "function" ) {
			return Promise.resolve(
				this.props.handleSubmit( this.state.inputValues, event ? event.target.name : null )
			)
				.then( alert => {
					if ( this.mounted ) {
						const stateChange = { ...initialState };
						if ( typeof alert === "string" && alert !== "" ) {
							stateChange.alert = alert;
						}

						this.setState( stateChange, this.runValidation );
					}
				} );
		}

		return null;
	}

	/**
	 * Handles events
	 * whenever a user clicks a button to reset the form's input data.
	 */
	handleReset( event = null ) {
		if ( event != null ) {
			event.preventDefault();
		}

		this.setState( { ...initialState }, () => this.runValidation() );
	}

	/**
	 * Composing output
	 */
	render() {
		if ( this.fieldList === null || typeof this.fieldList !== "object" ) {
			return <div className="empty-form"></div>;
		}

		const formFields = Object.keys( this.fieldList )
			.filter( name => [ "submit", "reset" ].includes( this.fieldList[name].type ) === false )
			.map( name => this.generateField( name ) )
			.filter( component => component != null );

		return React.createElement(
			Form,
			this.props.formClass ? { className: this.props.formClass } : {},
			[ this.generateAlert(), ...formFields, this.generateFinalRow() ]
		);
	}
}

FormGenerator.propTypes = propTypes;
FormGenerator.defaultProps = defaultProps;

export default FormGenerator;
