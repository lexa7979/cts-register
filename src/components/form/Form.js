/**
 * MIT License
 *
 * Copyright (c) 2019 <alexander.urban@cygni.se>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/* eslint-disable class-methods-use-this */

import assert from "assert";

import React from "react";
import PropTypes from "prop-types";
import { Form as ReactstrapForm, FormGroup, Col, Label, FormText } from "reactstrap";

import FormTypes from "./types";

const propTypes = {
	fields:       PropTypes.object.isRequired,
	handleSubmit: PropTypes.func.isRequired,
	formClass:    PropTypes.string,
	validateData: PropTypes.func,
};

const defaultProps = {};

/**
 * Component which contains an input form
 */
export class Form extends React.Component {
	/**
	 * Initialising component
	 */
	constructor( props ) {
		super( props );

		this.state = {
			inputValues: {},
			touched:     {},
			messages:    {},
			actions:     [],
		};

		this.fieldList = [];
		for ( const name in this.props.fields ) {
			if ( typeof name === "string" && name !== "" ) {
				const { type = "input", value = "" } = this.props.fields[name];
				assert(
					FormTypes[type] != null
					&& typeof FormTypes[type] === "object"
					&& typeof FormTypes[type].generateField === "function",
					`Unsupported field type "${type}" (${name})`
				);
				this.fieldList[name] = {
					...this.props.fields[name],

					name,
					type,
					id: `${this.props.formClass || "form"}-${name}`,
				};
				this.state.inputValues[name] = value;
			}
		}

		this.validation = {
			progressing: false,
			requested:   false,
		};

		this.handleInputChange = this.handleInputChange.bind( this );
		this.onSubmit = this.onSubmit.bind( this );
	}

	/**
	 * Actions to take when component is ready
	 */
	componentDidMount() {
		this.runValidation();
	}

	/**
	 * Validates the form inputs if props.validateData is given
	 *
	 * If another (asynchronous) validation is already in progress,
	 * nothing will be done, now. Still, it's memorised that
	 * another validation shall be run just after the active one
	 * has finished.
	 */
	runValidation() {
		if ( typeof this.props.validateData !== "function" ) {
			return Promise.resolve();
		}

		if ( this.validation.progressing ) {
			this.validation.requested = true;
			return Promise.resolve();
		}
		this.validation.progressing = true;

		return Promise.resolve( this.props.validateData( this.state.inputValues ) )
			.then( result => {
				this.validation.progressing = false;
				if ( this.validation.requested ) {
					this.validation.requested = false;
					this.runValidation();
					return;
				}

				if (
					result != null && typeof result === "object"
					&& Array.isArray( result.actions )
					&& result.messages != null && typeof result.messages === "object"
				) {
					this.setState( { actions: result.actions, messages: result.messages } );
				}
			} );
	}

	/**
	 * Wraps the given input component into a new div-container.
	 * A label-element will be added if requested.
	 */
	labelGenerator( fieldData, extraAttributes = {} ) {
		assert( fieldData != null && typeof fieldData === "object",
			"Invalid argument \"fieldData\"" );

		const { name, label, type, id } = fieldData;
		assert( typeof name === "string" && name !== "",
			`Invalid field property "name" (${name})` );
		assert( label == null || ( typeof label === "string" && label !== "" ),
			`Invalid field property "label" (${name}: ${label})` );
		assert( typeof type === "string" && type !== "",
			`Invalid field property "type" (${name}: ${type})` );
		assert( typeof id === "string" && id !== "",
			`Invalid field property "id" (${name}: ${id})` );

		const fieldProperties = {
			...extraAttributes,
			key:       `${name}-label`,
			className: `${label ? "label" : "nolabel"} ${type}`,
		};

		for ( const key in fieldProperties ) {
			if ( !fieldProperties[key] ) {
				delete fieldProperties[key];
			}
		}

		return label
			? React.createElement( Label, { ...fieldProperties, for: id }, label )
			: React.createElement( Col, fieldProperties );
	}

	/**
	 * If there is a message for the given input field,
	 * a matching component to show this text is composed.
	 *
	 * @param	{string}	inputName
	 */
	messageGenerator( inputName ) {
		assert( typeof inputName === "string" && inputName !== "", "Invalid argument" );

		return ( inputName === "submit" || this.state.touched[inputName] ) && this.state.messages[inputName]
			? <FormText key={`${inputName}-message`}>{this.state.messages[inputName]}</FormText>
			: null;
	}

	/**
	 *
	 * @param	{string}	name
	 */
	generateField( name ) {
		const data = this.fieldList[name];
		assert( data != null && typeof data === "object",
			`Invalid argument "name" (${name})` );
		assert( typeof data.type === "string" && typeof FormTypes[data.type] === "object",
			`Invalid field type (${data.type})` );

		const id = `${this.props.formClass || "form"}-${data.name}`;

		const label = this.labelGenerator( data, { sm: 3 } );
		const field = FormTypes[data.type].generateField.call( this, data, { id } );
		const message = this.messageGenerator( data.name );

		return <FormGroup row key={name}>
			{label}
			<Col sm={9}>
				{field}
				{message || ""}
			</Col>
		</FormGroup>;
	}

	/**
	 * Handles the change event of any form field
	 *
	 * @param	{object}	event
	 *		Description of the event to handle
	 */
	handleInputChange( event ) {
		const { name } = event.target;
		const value = event.target.type === "checkbox" ? event.target.checked : event.target.value;

		this.setState( { inputValues: { ...this.state.inputValues, [name]: value } }, () => {
			this.runValidation()
				.then( () => {
					if ( this.state.touched[name] !== true ) {
						this.setState( { touched: { ...this.state.touched, [name]: true } } );
					}
				} );
		} );
	}

	/**
	 * Handles events whenever a user clicks on one of the submit buttons
	 */
	onSubmit( event ) {
		event.preventDefault();

		if ( typeof this.props.handleSubmit === "function" ) {
			this.props.handleSubmit( this.state.inputValues );
		}
	}

	/**
	 * Composing output
	 */
	render() {
		if ( this.fieldList === null || typeof this.fieldList !== "object" ) {
			return <div className="empty-form"></div>;
		}

		const formGroups = [];
		const submitFields = [];
		for ( const name in this.fieldList ) {
			if ( typeof name === "string" && name !== "" ) {
				const data = this.fieldList[name];
				if ( data.type === "submit" ) {
					submitFields.push( FormTypes.submit.generateField.call( this, data ) );
				} else {
					const newField = this.generateField( name );
					if ( newField != null ) {
						formGroups.push( newField );
					}
				}
			}
		}

		if ( submitFields.length === 0 ) {
			submitFields.push( FormTypes.submit.generateField.call( this, { value: "Submit" } ) );
		}
		const submitMessage = this.messageGenerator( "submit" );

		return <ReactstrapForm
			className={this.props.formClass ? `form ${this.props.formClass}` : "form"}
			onSubmit={this.onSubmit}
		>
			{formGroups}
			<FormGroup row>
				<Col sm={{ size: 9, offset: 3 }}>
					{submitFields}
					{submitMessage}
				</Col>
			</FormGroup>
		</ReactstrapForm>;
	}
}

Form.propTypes = propTypes;
Form.defaultProps = defaultProps;

export default Form;
