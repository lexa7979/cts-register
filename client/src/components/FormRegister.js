/**
 * MIT License
 *
 * Copyright (c) 2019 <alexander.urban@cygni.se>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/* eslint-disable class-methods-use-this */

import React from "react";
import PropTypes from "prop-types";
import Axios from "axios";

import assert from "assert";

import { FormGenerator } from "./form";

const propTypes = {
	serverAvailable: PropTypes.bool,
};

const defaultProps = {};

/**
 * Component which will be used by users to register for the event.
 */
export class FormRegister extends React.Component {
	/**
	 * Initialising component
	 */
	constructor( props ) {
		super( props );

		this.formFields = {
			firstname: {
				type:        "input",
				label:       "Firstname:",
				placeholder: "Type in your firstname.",
				focus:       true,
			},
			lastname: {
				type:        "input",
				label:       "Lastname:",
				placeholder: "Type in your lastname.",
			},
			attending: {
				type:    "radio",
				label:   "Are you going to attend the conference?",
				options: { yes: "Yes", no: "No", maybe: "Maybe" },
			},
			submit: {
				type:  "submit",
				label: "Submit",
			},
			update: {
				type:  "submit",
				label: "Update",
			},
			reset: {
				type:  "reset",
				label: "Reset",
			},
		};

		this.validateData = this.validateData.bind( this );
	}

	/**
	 * Checks if the data of the input form is valid,
	 * also by contacting the database server.
	 *
	 * @param	{object}	inputData
	 *		Current data of the input form
	 *
	 * @returns	{Promise<object>}
	 *		Resolves with an object stating errors and possible submit-actions
	 */
	validateData( inputData ) {
		const messages = {
			firstname: inputData.firstname.trim() ? null : "Missing input: Firstname",
			lastname:  inputData.lastname.trim() ? null : "Missing input: Lastname",
			attending: inputData.attending ? null : "Choose one of the alternatives",
		};
		if ( Object.keys( messages ).filter( key => messages[key] != null ).length > 0 ) {
			return Promise.resolve( { actions: [ "reset" ], messages } );
		}

		if ( !this.props.serverAvailable ) {
			return Promise.resolve( { actions: [ "reset" ], messages: {} } );
		}

		return Axios.get(
			`/attendee/${inputData.firstname.trim()}/${inputData.lastname.trim()}`,
			{ validateStatus: null }
		)
			.then( response => {
				if ( response.status === 200 ) {
					const { success, item, code } = response.data;
					if ( success ) {
						if ( item.attending.toLowerCase() === inputData.attending.toLowerCase() ) {
							return {
								actions:  [ "reset" ],
								messages: { submit: "You answer was already saved, before." },
							};
						}
						return { actions: [ "update", "reset" ], messages: {} };
					}
					if ( code === "NOTFOUND" ) {
						return { actions: [ "submit", "reset" ], messages: {} };
					}
				}

				return {
					actions:  [ "reset" ],
					messages: { submit: response.data.error || response.data },
				};
			} );
	}

	/**
	 * Callback for class FormGenerator
	 * which will be used to store the entered user input.
	 *
	 * @param	{object}	inputData
	 * @param	{string}	submitFieldName
	 *
	 * @returns	{Promise<string>}
	 *		Resolves with a success-message which will be showed to the user
	 */
	handleSubmit( inputData, submitFieldName ) {
		return this.validateData( inputData )
			.then( result => {
				assert( Array.isArray( result.actions ) && result.actions.length > 0 );

				const data = {
					firstname: inputData.firstname,
					lastname:  inputData.lastname,
					attending: inputData.attending,
				};

				return Axios.put( "/attendee", data, { validateStatus: null } );
			} )
			.then( () => `Thank you, ${inputData.firstname}! `
				+ `You data was ${submitFieldName === "update" ? "updated" : "stored"} on the server.` );
	}

	/**
	 * Composing output
	 */
	render() {
		const disabled = this.props.serverAvailable
			? null
			: "Can't process your input, right now: No connection to database server";

		return <FormGenerator
			formClass="register"
			fields={this.formFields}
			handleSubmit={this.handleSubmit}
			validateData={this.validateData}
			disableWithMessage={disabled}
		/>;
	}
}

FormRegister.propTypes = propTypes;
FormRegister.defaultProps = defaultProps;

export default FormRegister;
