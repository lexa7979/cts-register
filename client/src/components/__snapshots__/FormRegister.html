
<!DOCTYPE html>
<html style="height: 100%;">
  <head>
  </head>
  <body style="display: flex; flex-flow: column nowrap; justify-content: center; align-items: center; height: 100%;">
    <h2>
      Component FormRegister - when rendering with no properties - delivers expected result  (-> check snapshot, too)
    </h2>
    <form class="register">
      <div class="row form-group">
        <label for="register-firstname"
               class="label input col-sm-3 col-form-label"
        >
          Firstname:
        </label>
        <div class="col-sm-9">
          <input type="text"
                 id="register-firstname"
                 name="firstname"
                 value
                 placeholder="Type in your firstname."
                 autofocus
                 class="form-control"
          >
        </div>
      </div>
      <div class="row form-group">
        <label for="register-lastname"
               class="label input col-sm-3 col-form-label"
        >
          Lastname:
        </label>
        <div class="col-sm-9">
          <input type="text"
                 id="register-lastname"
                 name="lastname"
                 value
                 placeholder="Type in your lastname."
                 class="form-control"
          >
        </div>
      </div>
      <div class="row form-group">
        <label for="register-attending"
               class="label radio col-sm-3 col-form-label"
        >
          Are you going to attend the conference?
        </label>
        <div class="col-sm-9">
          <div id="register-attending"
               class="col"
          >
            <div class="form-check form-check-inline">
              <label class="form-check-label">
                <input type="radio"
                       name="attending"
                       value="yes"
                       class="form-check-input"
                >
                Yes
              </label>
            </div>
            <div class="form-check form-check-inline">
              <label class="form-check-label">
                <input type="radio"
                       name="attending"
                       value="no"
                       class="form-check-input"
                >
                No
              </label>
            </div>
            <div class="form-check form-check-inline">
              <label class="form-check-label">
                <input type="radio"
                       name="attending"
                       value="maybe"
                       class="form-check-input"
                >
                Maybe
              </label>
            </div>
          </div>
        </div>
      </div>
      <div class="row form-group">
        <div class="col-sm-9 offset-sm-3">
          <button type="button"
                  name="submit"
                  value="Submit"
                  class="btn btn-primary"
          >
            Submit
          </button>
          <button type="button"
                  disabled
                  class="btn btn-secondary disabled"
          >
            Reset
          </button>
        </div>
      </div>
    </form>
    <br>
    <br>
    &lt;FormRegister /&gt;
  </body>
</html>
